### PR TITLE
completed lstm_cell

### PR DESCRIPTION
**Pull request by mei1127**
@huningxin  @fdwr @shiyi9801 @BruceDai  PTAL, thanks!

### CODE REVIEW

File: src/lstm_cell.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+ * @param {Tensor} input
+ * @param {Tensor} weight
+ * @param {Tensor} recurrentWeight
+ * @param {Tensor} hiddenState
+ * @param {Tensor} cellState
+ * @param {Number} hiddenSize
+ * @param {MLLstmCellOptions} options
+ * return {Tensor}
+ */
+
```
Comment (BruceDai) [inline src/lstm_cell.js:23]: Please remove this blank line.

File: test/lstm_cell_test.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
+'use strict';
+
+import {lstmCell} from '../src/lstm_cell.js';
+import {relu} from '../src/relu.js';
+import {Tensor} from '../src/lib/tensor.js';
+import * as utils from './utils.js';
+
+describe('test lstmCell', function() {
+  it.only('lstmCell lstmCell activations=[relu, relu, relu]', function() {
```
Comment (BruceDai) [inline test/lstm_cell_test.js:9]: Please use `it()` method.

File: test/lstm_cell_test.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    const batchSize = 2;
+    const inputSize = 2;
+    const hiddenSize = 2;
+    const input = new Tensor([batchSize, inputSize], [1, 2, 2, 1]);
+    const weight = new Tensor([4 * hiddenSize, inputSize],
+        new Float32Array([
```
Comment (BruceDai) [inline test/lstm_cell_test.js:15]: Prefer keep 2 space indentations.

**Comment by BruceDai** [comment]
Thanks @mei1127.
LGTM with some nits, please do some modifications.

File: src/lib/validate-input.js [hunk 4]
```diff
@@ -133,6 +133,73 @@ export function validateGemmParams(a, b) {
...
+  }
+  if (recurrentWeight.shape[0] !== 4 * hiddenSize || recurrentWeight.shape[1] !== hiddenSize) {
+    throw new Error(`The shape of recurrentWeight ` +
+    `[${recurrentWeight.shape[0]}, ${recurrentWeight.shape[1]}] is invalid.`);
+  }
+  if (hiddenState.rank !== 2) {
+    throw new Error(`The hiddenState (rank ${hiddenState.rank}) is not a 2-D tensor.`);
+  }
+  if (hiddenState.shape[0] !== batchSize || hiddenState.shape[1] !== hiddenSize) {
+    throw new Error(`The shape of hiddenState
```
Comment (fdwr) [inline src/lib/validate-input.js:164]: The error message is split.

**Review by fdwr — COMMENTED** [review]
Minor comments. Thanks Mei.

File: src/lstm_cell.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+      ),
+  );
+
+  // output cell state (ct)
+  const ct = add(mul(f, cellState), mul(i, g));
+
+  // output hidden state (ht)
+  const ht = mul(o, activation2(ct));
+
+  return [ht, ct];
```
[the remark this answers is not in the document]
Comment (Mei0211) [reply src/lstm_cell.js:144]: Maybe `hiddenStateNew` could be better for understand and differentiate from the previous state？

**Pull request merged**